core/src/main/java/io/methvin/watchservice/MacOSXListeningWatchService.java

```diff
@@ -114,21 +115,22 @@ public MacOSXListeningWatchService() {
   private final int kFSEventStreamCreateFlagFileEvents = 0x00000010;
 
   @Override
-  public AbstractWatchKey register(
+  public synchronized AbstractWatchKey register(
       WatchablePath watchable, Iterable<? extends WatchEvent.Kind<?>> events) throws IOException {
     checkOpen();
-    final MacOSXWatchKey watchKey = new MacOSXWatchKey(this, events, queueSize);
+    final MacOSXWatchKey watchKey = new MacOSXWatchKey(this, watchable, events, queueSize);
     final Path file = watchable.getFile().toAbsolutePath();
     // if we are already watching a parent of this directory, do nothing.
     for (Path watchedPath : pathsWatching) {
       if (file.startsWith(watchedPath)) return watchKey;
     }
+
     final Map<Path, HashCode> hashCodeMap = PathUtils.createHashCodeMap(file, fileHasher);
     final Pointer[] values = {CFStringRef.toCFString(file.toString()).getPointer()};
     final CFArrayRef pathsToWatch =
         CarbonAPI.INSTANCE.CFArrayCreate(null, values, CFIndex.valueOf(1), null);
-    final CarbonAPI.FSEventStreamCallback callback =
-        new MacOSXListeningCallback(watchKey, fileHasher, hashCodeMap);
+    final MacOSXListeningCallback callback =
+        new MacOSXListeningCallback(watchKey, fileHasher, hashCodeMap, file.toString(), file.toRealPath().toString());
```

Comment:
Any particular reason these need to be passed as strings versus as paths?

Reply:
done